File: src/main/java/org/example/ClassPrinter.java
package org.example;


import org.objectweb.asm.ClassVisitor;
import org.objectweb.asm.FieldVisitor;
import org.objectweb.asm.MethodVisitor;

import static org.objectweb.asm.Opcodes.ASM4;

public class ClassPrinter extends ClassVisitor {
    protected ClassPrinter() {
        super(ASM4);
    }

    @Override
    public void visit(int version, int access, String name, String signature, String superName, String[] interfaces) {
        super.visit(version, access, name, signature, superName, interfaces);
        System.out.println(name + " extends " + superName + " {");
    }


    @Override
    public FieldVisitor visitField(int access, String name, String descriptor, String signature, Object value) {
        System.out.println("   " + descriptor + " " + name);
        return null;
    }

    @Override
    public MethodVisitor visitMethod(int access, String name, String descriptor, String signature, String[] exceptions) {
        System.out.println("   " + descriptor + " " + name);
        return null;
    }

    @Override
    public void visitEnd() {
        super.visitEnd();
        System.out.println("}");
    }
}

File: build-logic/convention/src/main/kotlin/opclass/ClassPrinter.kt
package opclass

import org.objectweb.asm.*

/**
 * 打印Class内容
 */
class ClassPrinter(private val targetName: String) : ClassVisitor(Opcodes.ASM5) {

    companion object {
        const val TAG = "ClassPrinter"
    }

    override fun visit(
        version: Int,
        access: Int,
        name: String?,
        signature: String?,
        superName: String?,
        interfaces: Array<out String>?
    ) {
        if(name?.contains(targetName) == true) {
            println("class $name extends $superName implements ${interfaces.toString().let {
                it.subSequence(1, it.length - 1)
            }} {")
        } else {
            super.visit(version, access, name, signature, superName, interfaces)
        }
    }


    override fun visitField(
        access: Int,
        name: String?,
        descriptor: String?,
        signature: String?,
        value: Any?
    ): FieldVisitor? {
        println("   $access $descriptor $name = $value ;")
        return null
    }

    override fun visitMethod(
        access: Int,
        name: String?,
        descriptor: String?,
        signature: String?,
        exceptions: Array<out String>?
    ): MethodVisitor? {
        println("   $access $descriptor $name () throws $exceptions ; ")
        return null
    }

    override fun visitEnd() {
        println("}")
        super.visitEnd()
    }
}